ampproject/amp-wp
# In scripts expected to have JSON, ensure the JSON is valid

## Summary

* For `<script>` tags with `type="application/json"` or
`type="application/ld+json"`, this ensures that the `textContent` is valid JSON, and produces a `INVALID_JSON_CDATA` error if it's not.
* This doesn't help with #4318. It doesn't prevent `wp_staticize_emoji()` from making the emojis invalid. But it now reports emojis that have become invalid due to that issue.
* For example, adding a Custom HTML block with this will now cause an `INVALID_JSON_CDATA` error in the plugin:

```html
<amp-analytics><script type="application/ld+json">{"anEmoji": "🍫"}</script></amp-analytics>
```

## Testing
1. Add this to a plugin:
```php
add_action(
	'wp_head',
	static function() {
		?>
		<script type="application/ld+json">{"invalid": ⛷️}</script>
		<script type="application/ld+json">{"valid": "something"}</script>
		<script type="application/ld+json">{"anotherValid":" 🚫 🚫 🚫"}</script>
		<?php
	}
);

```
2. Create a new post, and add a Custom HTML block with this:

```html
<amp-analytics><script type="application/json">{"malformed": 🍫}</script>
```
3. Go to the front-end, and validate the URL
4. Expected: there are 2 validation errors:

![expected-actual](https://user-images.githubusercontent.com/4063887/75591565-a7a3e200-5a45-11ea-9d2e-e44d0b3a955c.png)

One should be from the first `<script>` in the `add_action( 'wp_head' ...` call above, and one should be from the Custom HTML block.

The bottom two scripts in the `add_action( 'wp_head' ...` call should not cause a validation error:

```html
<script type="application/ld+json">{"valid": "something"}</script>
<script type="application/ld+json">{"anotherValid":" 🚫 🚫 🚫"}</script>
```


Fixes #4320 

## Checklist

- [x] My pull request is addressing an [open issue](https://github.com/ampproject/amp-wp/contributing/project-management.md#life-of-an-issue) (please create one otherwise).
- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/contributing/engineering.md#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/contributing/engineering.md) (updates are often made to the guidelines, check it out periodically).
